src/LegalizeVectors.cpp

```diff
@@ -0,0 +1,630 @@
+#include "LegalizeVectors.h"
+#include "CSE.h"
+#include "Deinterleave.h"
+#include "DeviceInterface.h"
+#include "IRMutator.h"
+#include "IROperator.h"
+#include "Simplify.h"
+#include "Util.h"
+
+#include <optional>
+
+namespace Halide {
+namespace Internal {
+
+namespace {
+
+using namespace std;
+
+int max_lanes_for_device(DeviceAPI api, int parent_max_lanes) {
+    switch (api) {
+    case DeviceAPI::Metal:
+    case DeviceAPI::WebGPU:
+    case DeviceAPI::Vulkan:
+    case DeviceAPI::D3D12Compute:
+        return 4;
+    case DeviceAPI::OpenCL:
+        return 16;
+    case DeviceAPI::CUDA:
+    case DeviceAPI::Hexagon:
+    case DeviceAPI::HexagonDma:
+    case DeviceAPI::Host:
+        return 0;  // No max: LLVM based legalization
+    case DeviceAPI::None:
+        return parent_max_lanes;
+    case DeviceAPI::Default_GPU:
+        internal_error << "No GPU API was selected.";
+        return 0;
+    }
+    internal_error << "Unknown Device API";
+    return 0;
+}
+
+std::string vec_name(const string &name, int lane_start, int lane_count) {
+    return name + ".lanes_" + std::to_string(lane_start) + "_" + std::to_string(lane_start + lane_count - 1);
+}
+
+Expr simplify_shuffle(const Shuffle *op) {
```

Comment:
Why is this here rather than in the simplifier?

Reply:
I'll try this tomorrow. It indeed is late.